src/Microsoft.AspNet.Routing/Template/TemplateBinder.cs

```diff
@@ -471,13 +472,24 @@ public bool Accept(string value)
                 BufferState = SegmentState.Inside;
                 UriState = SegmentState.Inside;
 
-                _uri.Append(value);
+                // Allow the first segment to have a leading slash.
+                // This prevents the leading slash from PathString segments from being encoded.
+                if (_uri.Length == 0 && value.Length > 0 && value[0] == '/')
+                {
+                    _uri.Append("/");
+                    _urlEncoder.Encode(_uriWriter, value, 1, value.Length - 1);
+                }
+                else
+                {
+                    _urlEncoder.Encode(_uriWriter, value);
+                }
+
                 return true;
             }
 
             public void Remove(string literal)
             {
-                _uri.Length -= literal.Length;
+                _uri.Length -= _urlEncoder.Encode(literal).Length;
```

Comment:
I think we're almost always removing the last segment we accepted. We should be able to keep track of how long the encoded length was.